lms/djangoapps/courseware/entrance_exams.py

```diff
@@ -0,0 +1,157 @@
+    Use this operation instead of checking the feature flag all over the place
+    """
+    return settings.FEATURES.get('ENTRANCE_EXAMS', False)
+
+
+def course_has_entrance_exam(course):
+    """
+    Checks to see if a course is properly configured for an entrance exam
+    """
+    if not feature_is_enabled():
+        return False
+    if not course.entrance_exam_enabled:
+        return False
+    if not course.entrance_exam_id:
+        return False
+    return True
+
+
+def user_can_skip_entrance_exam(request, user, course):
+    """
+    Checks all of the various override conditions for a user to skip an entrance exam
+    """
+    if not course_has_entrance_exam(course):
+        return True
+    if user.is_anonymous():
+        return False
+    if has_access(user, 'staff', course):
+        return True
+    if EntranceExamConfiguration.user_can_skip_entrance_exam(user, course.id):
+        return True
+    exam_content = get_entrance_exam_content(request, course)
+    if not exam_content:
+        return True
+    return False
+
+
+def user_has_passed_entrance_exam(request, user, course):
+    """
+    Checks to see if the user has attained a sufficient score to pass the exam
+    """
+    if not course_has_entrance_exam(course):
+        return True
+    if user.is_anonymous():
+        return False
+    entrance_exam_score = get_entrance_exam_score(request, course)
+    if entrance_exam_score >= course.entrance_exam_minimum_score_pct:
+        return True
+    return False
+
+
+# pylint: disable=invalid-name
+def user_can_access_courseware_with_entrance_exam(request, user, course):
+    """
+    Checks all of the various access/override workflows to see if the user
+    is allowed to access the course content, or if they have to view the exam
+    """
+    if not course_has_entrance_exam(course):
+        return True
+    if user_can_skip_entrance_exam(request, user, course):
+        return True
+    if user_has_passed_entrance_exam(request, user, course):
+        return True
+    return False
+
+
+def _calculate_entrance_exam_score(user, course_descriptor, exam_modules):
+    """
+    Calculates the score (percent) of the entrance exam using the provided modules
+    """
+    # All of the exam module ids
+    exam_module_ids = [exam_module.location for exam_module in exam_modules]
+
+    # All of the corresponding student module records
+    student_modules = StudentModule.objects.filter(
+        student=user,
+        course_id=course_descriptor.id,
+        module_state_key__in=exam_module_ids,
+    )
+    exam_pct = 0
+    module_pcts = []
+    ignore_categories = ['course', 'chapter', 'sequential', 'vertical']
+
+    for module in exam_modules:
+        if module.graded and module.category not in ignore_categories:
+            module_pct = 0
+            for student_module in student_modules:
+                if unicode(student_module.module_state_key) == unicode(module.location) and student_module.max_grade:
```

Comment:
Instead of iterating through student_modules every time we're looking at a new exam_module, how about converting student_modules from a queryset to a dict mapping module_state_keys to their corresponding student_modules?